core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java

```diff
@@ -83,83 +84,200 @@ public static SortedSet<String> listSchemas(Session session, Metadata metadata,
 
     public static SortedSet<String> listSchemas(Session session, Metadata metadata, AccessControl accessControl, String catalogName, Optional<String> schemaName)
     {
-        Set<String> schemaNames = ImmutableSet.copyOf(metadata.listSchemaNames(session, catalogName));
-        if (schemaName.isPresent()) {
-            // we don't use metadata.schemaExists(), because this would change semantics of the method (all vs visible schemas)
-            if (!schemaNames.contains(schemaName.get())) {
-                return ImmutableSortedSet.of();
+        try {
+            Set<String> schemaNames = ImmutableSet.copyOf(metadata.listSchemaNames(session, catalogName));
+            if (schemaName.isPresent()) {
+                // we don't use metadata.schemaExists(), because this would change semantics of the method (all vs visible schemas)
+                if (!schemaNames.contains(schemaName.get())) {
+                    return ImmutableSortedSet.of();
+                }
+                schemaNames = ImmutableSet.of(schemaName.get());
             }
-            schemaNames = ImmutableSet.of(schemaName.get());
+            return ImmutableSortedSet.copyOf(accessControl.filterSchemas(session.toSecurityContext(), catalogName, schemaNames));
+        }
+        catch (RuntimeException exception) {
+            if (exception instanceof TrinoException trinoException) {
+                throw new TrinoException(
+                        trinoException::getErrorCode,
+                        "Error listing schema names for catalog %s: %s".formatted(catalogName, exception.getMessage()),
+                        exception);
+            }
+            throw new TrinoException(
+                    StandardErrorCode.GENERIC_INTERNAL_ERROR,
+                    "Error listing schema names for catalog %s".formatted(catalogName),
```

Comment:
You don't need to use `formatted` when you are adding thin at the end. `"Error ... for catalog: " + catalog`